velox/expression/CastExpr.cpp

```diff
@@ -520,63 +520,73 @@ VectorPtr CastExpr::applyDecimal(
-template <bool castTo>
-void applyCustomTypeCast(
-    VectorPtr& input,
-    DecodedVector& inputDecoded,
-    const SelectivityVector& allRows,
-    const SelectivityVector& nonNullRows,
-    const CastOperatorPtr& castOperator,
-    const TypePtr& thisType,
-    const TypePtr& otherType,
+void CastExpr::applyComplexType(
+    const SelectivityVector& rows,
+    const BaseVector& input,
     exec::EvalCtx& context,
-    bool nullOnFailure,
+    const TypePtr& fromType,
+    const TypePtr& toType,
     VectorPtr& result) {
-  VELOX_CHECK_NE(
-      thisType,
-      otherType,
-      "Attempting to cast from {} to itself.",
-      thisType->toString());
-
-  exec::LocalSelectivityVector baseRows(
-      *context.execCtx(), inputDecoded.base()->size());
-  baseRows->clearAll();
-  context.applyToSelectedNoThrow(nonNullRows, [&](auto row) {
-    baseRows->setValid(inputDecoded.index(row), true);
-  });
-  baseRows->updateBounds();
+  switch (toType->kind()) {
+    // Handle complex type conversions
+    case TypeKind::MAP:
+      result = applyMap(
+          rows,
+          input.asUnchecked<MapVector>(),
+          context,
+          fromType->asMap(),
+          toType->asMap());
+      break;
+    case TypeKind::ARRAY:
+      result = applyArray(
+          rows,
+          input.asUnchecked<ArrayVector>(),
+          context,
+          fromType->asArray(),
+          toType->asArray());
+      break;
+    case TypeKind::ROW:
+      result = applyRow(
+          rows,
+          input.asUnchecked<RowVector>(),
+          context,
+          fromType->asRow(),
+          toType->asRow());
+      break;
+    case TypeKind::SHORT_DECIMAL:
+    case TypeKind::LONG_DECIMAL:
+      result = applyDecimal(rows, input, context, fromType, toType);
+      break;
+    default: {
+      VELOX_UNREACHABLE();
+    }
+  }
+}
 
-  VectorPtr localResult;
-  if constexpr (castTo) {
-    context.ensureWritable(*baseRows, thisType, localResult);
+void CastExpr::applyComplexAndCustomType(
+    const SelectivityVector& rows,
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    const TypePtr& fromType,
+    const TypePtr& toType,
+    VectorPtr& result) {
+  if (castFromOperator_ || castToOperator_) {
+    VELOX_CHECK_NE(
+        fromType,
+        toType,
+        "Attempting to cast from {} to itself.",
+        fromType->toString());
 
-    castOperator->castTo(
-        *inputDecoded.base(), context, *baseRows, nullOnFailure, *localResult);
+    if (castToOperator_) {
+      castToOperator_->castTo(
+          input, context, rows, nullOnFailure_, toType, result);
+    } else {
+      castFromOperator_->castFrom(
+          input, context, rows, nullOnFailure_, toType, result);
+    }
   } else {
-    context.ensureWritable(*baseRows, otherType, localResult);
-
-    castOperator->castFrom(
-        *inputDecoded.base(), context, *baseRows, nullOnFailure, *localResult);
+    applyComplexType(rows, input, context, fromType, toType, result);
```

Comment:
applyComplexType method seems to be used only once. Consider folding it into the only caller.